daemon/rrd-update.c

```diff
@@ -243,34 +247,69 @@ void rb_rrd_update(rb_poller *poll)
                 log_errorx("couldn't create directory for raw file: %s : %s",
                             path,  strerror(errno));
                 free(parent);
-                /* Try the next item */
-                continue;
+                break; /* next raw file */
             }
             free(parent);
-        
-            fp = fopen(path, "a");
-            if(fp == NULL)
-            {
-                log_errorx("couldn't open raw file: %s for writing : %s",
-                            path,  strerror(errno));
-                /* Try the next item */
-                continue;
+
+            fd = open(path, O_WRONLY|O_APPEND|O_CREAT);
+            if(fd == -1) {
+                log_errorx("open raw file for append failed: %s: %s", path, strerror(errno));
+                break; /* next raw file */
             }
 
-            /* Write item record to raw file */
-            fprintf(fp, "%" PRId64 "\t", time);
-            if(item->reference)
-                fprintf(fp, "%s", item->reference);
-            else
-                fprintf(fp, "%s", item->field);
-            fprintf(fp, "\t");
-            if(item->vtype == VALUE_REAL)
-                fprintf(fp, "%" PRId64, item->v.i_value);
-            else if(item->vtype == VALUE_FLOAT)
-                fprintf(fp, "%.4lf", item->v.f_value);
-            fprintf(fp, "\n");
-
-            fclose(fp);
+	    write_sample(fd, path, &time, item);
+
+	    if (close(fd) == -1) {
+		    log_errorx("close of raw file failed: %s: %s", path, strerror(errno));
+	    }
         }
     }
 }
+
+static void
+write_sample(int fd, const char* path, const time_t *time, const rb_item *item)
+{
+	char buf[RAW_BUFLEN];
+	int n;
+
+	switch (item->vtype) {
+	case VALUE_REAL:
+		n = snprintf(buf, sizeof(buf), "%"PRId64"\t%s\t%"PRId64"\n",
+		    *time,
+		    (item->reference ? item->reference : item->field),
+		    item->v.i_value);
+		break;
+
+	case VALUE_FLOAT:
+		n = snprintf(buf, sizeof(buf), "%"PRId64"\t%s\t%.4lf\n",
+		    *time,
+		    (item->reference ? item->reference : item->field),
+		    item->v.f_value);
+		break;
+
+	case VALUE_UNSET:
+		n = snprintf(buf, sizeof(buf), "%"PRId64"\t%s\n",
+		    *time,
+		    (item->reference ? item->reference : item->field));
+		break;
+
+	default:
+		log_errorx("raw file: %s: unknown sample value type: %d", path, item->vtype);
+		return;
+	}
+
+	if (n == -1) {
+		log_errorx("raw file: %s: snprintf: %s", path, strerror(errno));
+		return;
+	}
+
+	if (n >= sizeof(buf)) {
+		log_errorx("raw file: %s: truncated sample string: required: %d", path, n);
+		return;
+	}
+
+	if (write(fd, buf, strlen(buf)) == -1) {
```

Comment:
Please check the number of bytes written by write, not just compare to -1:

man-page for write(2) says:
    RETURN VALUE
       On success, the number of bytes written is returned (zero indicates nothing was written).  It is not an  error
       if  this  number  is  smaller than the number of bytes requested; this may happen for example because the disk
       device was filled.  See also NOTES.

I assume we do not want this situation ignored...